### PR TITLE
Bug 1557828 - Record error on too many list items

### PR DESCRIPTION
**Pull request by badboy**
Followup for #129.

Also see commit messages for details.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/src/metrics/string_list.rs [hunk 1]
```diff
@@ -66,25 +66,29 @@ impl StringListMetric {
...
-                        //);
-                        // TODO (bug 1557828) - record the error when it doesn't deadlock.
-                        // record_error(glean, &self.meta, ErrorType::InvalidValue, msg);
+                        let msg = format!(
+                            "String list length of {} exceeds maximum of {}",
+                            old_value.len() + 1,
+                            MAX_LIST_LENGTH
+                        );
+                        error = Some(msg);
                         return Metric::StringList(old_value);
```
Comment (mdboom) [inline glean-core/src/metrics/string_list.rs]: I find the early return inside an anonymous function a bit confusing.

Maybe move the `old_value.push(value.clone())` into an `else` clause, but always return `Metric::StringList(old_value)` in the same place at the end of the expression?

File: glean-core/tests/string_list.rs [hunk 2]
```diff
@@ -195,7 +195,10 @@ fn string_lists_dont_exceed_max_items() {
 
     // TODO (bug 1557828) - Uncomment when the error starts being recorded again.
```
Comment (Dexterp37) [inline glean-core/tests/string_list.rs]: nit: let's remove the comment here

File: glean-core/tests/string_list.rs [hunk 3]
```diff
@@ -211,7 +214,7 @@ fn string_lists_dont_exceed_max_items() {
 
     // TODO (bug 1557828) - Increment by 1 when the previous error starts being recorded again.
```
Comment (Dexterp37) [inline glean-core/tests/string_list.rs]: nit: let's remove the comment here

**Comment by badboy** [comment]
> If I understand this correctly we're forced to use a mut transform because that's the only way we can get the error out?

Yes, before we took an `Fn`, which is a closure without an environment, which therefore it cannot read or write to anything outside of it.
The storage doesn't really care about what the closure does, so it's fine relaxing the bound to `FnMut` which allows the closure to reference stuff on its outside (by actually taking references to stuff outside).

**Pull request merged**